File: kosmo-orbit/packages/kosmo-lizenz/src/lizenz.ts
/**
 * @kosmo/lizenz — signierte Lizenz, reine Verify-Funktion (Serie I / Batch B6).
 *
 * **Ehrliches Urteil (docs/SERIE-I-BUILDPLAN.md §3):** eine rein clientseitige
 * Lizenzprüfung ist grundsätzlich umgehbar — der öffentliche Schlüssel steckt
 * im Build und ist patchbar, ein entschlossener Angreifer entfernt den Check.
 * Diese Prüfung ist deshalb bewusst **Reibung, keine Garantie**: sie blendet
 * einen ehrlichen Hinweis ein, sperrt aber NIE die lokale Arbeit hart aus.
 * Der einzige wirklich harte Hebel ist die **Server-Bindung** (Sync-`onAuthenticate`
 * / Bridge-`token_guard`) — die läuft server-seitig, wo ein Client-Patch nicht
 * hinreicht.
 *
 * **Kein privater Schlüssel in diesem Paket.** Nur `verifiziereLizenz` (prüft
 * eine fertige, signierte Lizenz gegen einen ÖFFENTLICHEN Schlüssel) lebt hier.
 * Das Signieren selbst passiert auf Owner-Infrastruktur; ein Test-Keypair für
 * die Unit-Tests liegt separat unter `test/test-hilfen.ts` — nie hier im
 * Haupt-Modul, damit kein Signierwerkzeug versehentlich mitgebaut wird.
 *
 * **Web Crypto statt npm-Dep**: `crypto.subtle` mit Ed25519 läuft nativ in
 * Node ≥ 20 und in allen modernen Browsern — kein zusätzliches Paket, keine
 * Lieferketten-Fläche.
 *
 * `jetzt` ist immer ein Parameter, nie `Date.now()`/`new Date()` intern —
 * damit `verifiziereLizenz` deterministisch testbar bleibt (gültig/abgelaufen
 * lässt sich mit einem festen Testdatum exakt nachstellen).
 */

/** Lizenz-Datenmodell — die Felder, die tatsächlich signiert werden. */
export interface LizenzDaten {
  /** Wer die Lizenz hält (Büro-/Personenname). */
  inhaber: string;
  /** Edition/Stufe, z.B. `standard`, `remote`, `cloud`. */
  edition: string;
  /** ISO-Datum (`YYYY-MM-DD`) oder voller ISO-Zeitstempel — gültig bis Ende dieses Tages/Zeitpunkts. */
  gueltigBis: string;
  /** ISO-Zeitstempel der Ausstellung (Dokumentation, fliesst nicht in die Gültigkeitsprüfung ein). */
  ausgestelltAm: string;
  /** Eindeutige Lizenz-ID — Grundlage der Widerrufsliste. */
  lizenzId: string;
}

/** Das signierte Paket: Daten + Ed25519-Signatur (base64) über die kanonische Form. */
export interface LizenzPaket {
  daten: LizenzDaten;
  /** base64-kodierte Ed25519-Signatur (64 Rohbytes) über `kanonischeLizenznachricht(daten)`. */
  signatur: string;
}

export type LizenzGrund =
  | 'lizenztext_ungueltig'
  | 'lizenzdaten_unvollstaendig'
  | 'oeffentlicher_schluessel_ungueltig'
  | 'signaturformat_ungueltig'
  | 'signatur_ungueltig'
  | 'abgelaufen'
  | 'widerrufen';

export interface LizenzErgebnis {
  gueltig: boolean;
  /** Nur gesetzt, wenn `gueltig === false`. */
  grund?: LizenzGrund;
  /** Gesetzt, sobald die Daten lesbar sind — auch bei `abgelaufen`/`widerrufen`, damit die App den Inhaber trotzdem anzeigen kann. */
  lizenz?: LizenzDaten;
}

const PFLICHTFELDER: (keyof LizenzDaten)[] = [
  'inhaber',
  'edition',
  'gueltigBis',
  'ausgestelltAm',
  'lizenzId',
];

/**
 * Kanonische, sprachübergreifend reproduzierbare Signier-Nachricht — eine
 * feste Feldreihenfolge statt JSON-Schlüsselreihenfolge (die zwischen
 * TypeScript und der Python-Implementierung in der Bridge sonst driften
 * könnte). Jede Änderung hier bricht ALLE bestehenden Lizenzen — nicht ohne
 * Migrationsplan anfassen.
 */
export function kanonischeLizenznachricht(daten: LizenzDaten): string {
  return [
    'kosmo-lizenz/v1',
    `inhaber=${daten.inhaber}`,
    `edition=${daten.edition}`,
    `gueltigBis=${daten.gueltigBis}`,
    `ausgestelltAm=${daten.ausgestelltAm}`,
    `lizenzId=${daten.lizenzId}`,
  ].join('\n');
}

/** UTF-8-sichere Bytes → base64 (kein `Buffer` — läuft im Browser wie in Node). */
export function bytesZuBase64(bytes: Uint8Array): string {
  let bin = '';
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]!);
  return btoa(bin);
}

/** base64 → Bytes (kein `Buffer`). Wirft bei kaputtem base64. */
export function base64ZuBytes(b64: string): Uint8Array {
  const bin = atob(b64);
  const bytes = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
  return bytes;
}

/** Lesbare `LizenzDaten`? Nur einfache String-Prüfung — kein zod nötig für 5 Felder. */
function datenVollstaendig(wert: unknown): wert is LizenzDaten {
  if (!wert || typeof wert !== 'object') return false;
  const d = wert as Record<string, unknown>;
  return PFLICHTFELDER.every((feld) => typeof d[feld] === 'string' && d[feld]!.trim().length > 0);
}

/** `gueltigBis` → Zeitpunkt, bis zu dem die Lizenz gilt (Ende des Tages bei reinem Datum). */
function gueltigBisMs(gueltigBis: string): number {
  const nurDatum = /^\d{4}-\d{2}-\d{2}$/.test(gueltigBis);
  return Date.parse(nurDatum ? `${gueltigBis}T23:59:59.999Z` : gueltigBis);
}

/**
 * Baut das Lizenz-Text-Format: base64(UTF-8(JSON(paket))) — ein einziger
 * opaker String, der sich in ein Textfeld einfügen lässt (kein `Buffer`,
 * kein JSON mit Zeilenumbrüchen im Klartext).
 */
export function kodiereLizenztext(paket: LizenzPaket): string {
  const json = JSON.stringify(paket);
  return bytesZuBase64(new TextEncoder().encode(json));
}

function dekodiereLizenztext(lizenzText: string): LizenzPaket | null {
  try {
    const json = new TextDecoder().decode(base64ZuBytes(lizenzText));
    const paket = JSON.parse(json) as unknown;
    if (
      paket &&
      typeof paket === 'object' &&
      'daten' in paket &&
      'signatur' in paket &&
      typeof (paket as { signatur: unknown }).signatur === 'string'
    ) {
      return paket as LizenzPaket;
    }
    return null;
  } catch {
    return null;
  }
}

/**
 * Prüft eine signierte Lizenz gegen einen öffentlichen Ed25519-Schlüssel.
 *
 * @param lizenzText  base64(UTF-8(JSON({daten, signatur}))) — siehe `kodiereLizenztext`.
 * @param publicKeyBase64  32 Roh-Bytes (raw, base64) — der ÖFFENTLICHE Schlüssel aus dem Build/Env.
 * @param jetzt  der Prüfzeitpunkt — niemals intern `new Date()`, damit der Test deterministisch bleibt.
 */
export async function verifiziereLizenz(
  lizenzText: string,
  publicKeyBase64: string,
  jetzt: Date,
): Promise<LizenzErgebnis> {
  const paket = dekodiereLizenztext(lizenzText);
  if (!paket) return { gueltig: false, grund: 'lizenztext_ungueltig' };
  if (!datenVollstaendig(paket.daten)) {
    return { gueltig: false, grund: 'lizenzdaten_unvollstaendig' };
  }

  let pubBytes: Uint8Array;
  let sigBytes: Uint8Array;
  try {
    pubBytes = base64ZuBytes(publicKeyBase64);
    if (pubBytes.length !== 32) throw new Error('falsche Länge');
  } catch {
    return { gueltig: false, grund: 'oeffentlicher_schluessel_ungueltig' };
  }
  try {
    sigBytes = base64ZuBytes(paket.signatur);
    if (sigBytes.length !== 64) throw new Error('falsche Länge');
  } catch {
    return { gueltig: false, grund: 'signaturformat_ungueltig' };
  }

  try {
    const key = await crypto.subtle.importKey('raw', pubBytes as BufferSource, { name: 'Ed25519' }, false, [
      'verify',
    ]);
    const nachricht = new TextEncoder().encode(kanonischeLizenznachricht(paket.daten));
    const ok = await crypto.subtle.verify('Ed25519', key, sigBytes as BufferSource, nachricht as BufferSource);
    if (!ok) return { gueltig: false, grund: 'signatur_ungueltig', lizenz: paket.daten };
  } catch {
    return { gueltig: false, grund: 'signatur_ungueltig' };
  }

  if (jetzt.getTime() > gueltigBisMs(paket.daten.gueltigBis)) {
    return { gueltig: false, grund: 'abgelaufen', lizenz: paket.daten };
  }

  return { gueltig: true, lizenz: paket.daten };
}

/** Reine Widerrufsprüfung — getrennt von `verifiziereLizenz`, weil Server
 * (Sync/Bridge) die Liste selbst laden (Env/Datei), das Verify-Ergebnis aber
 * unverändert lassen sollen. */
export function istWiderrufen(lizenzId: string, widerrufsliste: readonly string[]): boolean {
  return widerrufsliste.includes(lizenzId);
}

File: kosmo-orbit/tools/homestation-bridge/kosmo_bridge/lizenz.py
"""Signierte Lizenz — Python-Verify für die Bridge (Serie I / Batch B6).

Getrennt von `main.py`, damit die reine Entscheidungslogik ohne FastAPI/
Uvicorn-Start unit-testbar bleibt. Muss BYTE-FÜR-BYTE dieselbe kanonische
Nachricht und dasselbe Lizenz-Text-Format bauen wie `@kosmo/lizenz`
(TypeScript, Web Crypto) — sonst würde eine mit dem TS-Signierwerkzeug
ausgestellte Lizenz hier nicht verifizieren. Format:

  lizenzText = base64( utf8( json.dumps({"daten": {...}, "signatur": "<b64>"}) ) )
  Signatur   = Ed25519(kanonische_nachricht(daten)) über:
    "kosmo-lizenz/v1\\ninhaber=...\\nedition=...\\ngueltigBis=...\\nausgestelltAm=...\\nlizenzId=..."

**Kein privater Schlüssel hier oder sonst irgendwo im Repo.** Diese Datei
prüft nur gegen einen ÖFFENTLICHEN Schlüssel (Env `KOSMO_BRIDGE_LIZENZ_PUBKEY`).

**Ed25519-Bibliothek**: nutzt `cryptography`, wenn installiert, sonst
`PyNaCl`, sonst keine — dann lehnt `lizenz_pruefen` bei aktiver Pflicht JEDE
Lizenz ehrlich ab (`grund="keine_ed25519_lib"`), statt eine Prüfung
vorzutäuschen, die gar nicht laufen kann (fail closed, kein Fake-Schutz).
"""

from __future__ import annotations

import base64
import json
import re
from datetime import datetime

ED25519_LIB: str | None
try:
    from cryptography.exceptions import InvalidSignature
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

    ED25519_LIB = "cryptography"
except ImportError:
    try:
        import nacl.exceptions
        import nacl.signing

        ED25519_LIB = "pynacl"
    except ImportError:
        ED25519_LIB = None

LIZENZ_PFLICHTFELDER = ("inhaber", "edition", "gueltigBis", "ausgestelltAm", "lizenzId")


def kanonische_lizenznachricht(daten: dict) -> bytes:
    """Identische Feldreihenfolge wie `kanonischeLizenznachricht()` in
    `@kosmo/lizenz` (TypeScript) — nicht ohne Migrationsplan ändern."""
    zeilen = ["kosmo-lizenz/v1"] + [f"{feld}={daten.get(feld, '')}" for feld in LIZENZ_PFLICHTFELDER]
    return "\n".join(zeilen).encode("utf-8")


def _signatur_gueltig(nachricht: bytes, signatur: bytes, oeffentlicher_schluessel: bytes) -> bool:
    if ED25519_LIB == "cryptography":
        try:
            Ed25519PublicKey.from_public_bytes(oeffentlicher_schluessel).verify(signatur, nachricht)
            return True
        except (InvalidSignature, ValueError, TypeError):
            return False
    if ED25519_LIB == "pynacl":
        try:
            nacl.signing.VerifyKey(oeffentlicher_schluessel).verify(nachricht, signatur)
            return True
        except (nacl.exceptions.BadSignatureError, ValueError, TypeError):
            return False
    return False  # keine Ed25519-Lib verfügbar — ehrlich verweigern


def _gueltig_bis_ms(gueltig_bis: str) -> float:
    """`gueltigBis` → Zeitpunkt (ms seit Epoch), bis zu dem die Lizenz gilt.
    Ein reines Datum (`YYYY-MM-DD`) gilt bis zum Ende dieses Tages (UTC)."""
    nur_datum = bool(re.fullmatch(r"\d{4}-\d{2}-\d{2}", gueltig_bis))
    iso = f"{gueltig_bis}T23:59:59.999+00:00" if nur_datum else gueltig_bis.replace("Z", "+00:00")
    return datetime.fromisoformat(iso).timestamp() * 1000


def _daten_vollstaendig(daten: object) -> bool:
    if not isinstance(daten, dict):
        return False
    return all(isinstance(daten.get(f), str) and daten.get(f, "").strip() for f in LIZENZ_PFLICHTFELDER)


def _dekodiere_lizenztext(lizenz_text: str) -> dict | None:
    try:
        json_bytes = base64.b64decode(lizenz_text, validate=True)
        paket = json.loads(json_bytes.decode("utf-8"))
        if not isinstance(paket, dict) or "daten" not in paket or "signatur" not in paket:
            return None
        if not isinstance(paket["signatur"], str):
            return None
        return paket
    except Exception:
        return None


def lizenz_pruefen(lizenz_text: str, public_key_b64: str, jetzt: datetime) -> dict:
    """Prüft eine signierte Lizenz gegen einen öffentlichen Ed25519-Schlüssel.

    Rückgabe: `{"gueltig": bool, "grund": str | None, "lizenz": dict | None}`.
    `jetzt` ist immer ein Parameter — nie `datetime.now()` intern — damit die
    Funktion deterministisch testbar bleibt (gültig/abgelaufen mit festem
    Testdatum exakt nachstellbar, wie in `@kosmo/lizenz`).
    """
    paket = _dekodiere_lizenztext(lizenz_text)
    if paket is None:
        return {"gueltig": False, "grund": "lizenztext_ungueltig", "lizenz": None}

    daten = paket.get("daten")
    if not _daten_vollstaendig(daten):
        return {"gueltig": False, "grund": "lizenzdaten_unvollstaendig", "lizenz": None}

    try:
        pubkey_bytes = base64.b64decode(public_key_b64, validate=True)
        if len(pubkey_bytes) != 32:
            raise ValueError("falsche Länge")
    except Exception:
        return {"gueltig": False, "grund": "oeffentlicher_schluessel_ungueltig", "lizenz": None}

    try:
        sig_bytes = base64.b64decode(paket["signatur"], validate=True)
        if len(sig_bytes) != 64:
            raise ValueError("falsche Länge")
    except Exception:
        return {"gueltig": False, "grund": "signaturformat_ungueltig", "lizenz": None}

    if ED25519_LIB is None:
        return {"gueltig": False, "grund": "keine_ed25519_lib", "lizenz": None}

    nachricht = kanonische_lizenznachricht(daten)
    if not _signatur_gueltig(nachricht, sig_bytes, pubkey_bytes):
        return {"gueltig": False, "grund": "signatur_ungueltig", "lizenz": daten}

    if jetzt.timestamp() * 1000 > _gueltig_bis_ms(daten["gueltigBis"]):
        return {"gueltig": False, "grund": "abgelaufen", "lizenz": daten}

    return {"gueltig": True, "grund": None, "lizenz": daten}


def lade_widerrufsliste(env: dict) -> list[str]:
    """Widerrufsliste aus Env-Kommaliste + optionaler Datei (eine Lizenz-ID
    pro Zeile ODER JSON-Array) — analog zu `ladeWiderrufsliste()` im
    Sync-Server. Fehlt/kaputt die Datei, wird sie ehrlich ignoriert."""
    ids: set[str] = set()
    for teil in str(env.get("KOSMO_BRIDGE_LIZENZ_WIDERRUF", "")).split(","):
        t = teil.strip()
        if t:
            ids.add(t)
    datei = env.get("KOSMO_BRIDGE_LIZENZ_WIDERRUF_DATEI")
    if datei:
        try:
            from pathlib import Path

            inhalt = Path(datei).read_text(encoding="utf-8").strip()
            werte = json.loads(inhalt) if inhalt.startswith("[") else inhalt.splitlines()
            for w in werte:
                t = str(w).strip()
                if t:
                    ids.add(t)
        except Exception:
            pass
    return list(ids)


def ist_widerrufen(lizenz_id: str, widerrufsliste: list[str]) -> bool:
    return lizenz_id in widerrufsliste
